libimage/image.go

```diff
@@ -708,17 +710,44 @@ func (i *Image) HasDifferentDigest(ctx context.Context, remoteRef types.ImageRef
 		return false, err
 	}
 
-	rawManifest, _, err := remoteImg.Manifest(ctx)
+	rawManifest, rawManifestMIMEType, err := remoteImg.Manifest(ctx)
 	if err != nil {
 		return false, err
 	}
 
-	remoteDigest, err := manifest.Digest(rawManifest)
-	if err != nil {
-		return false, err
+	// If the remote ref's manifest is a list, try to zero in on the image
+	// in the list that we would eventually try to pull.
+	var remoteDigest digest.Digest
+	if manifest.MIMETypeIsMultiImage(rawManifestMIMEType) {
+		list, err := manifest.ListFromBlob(rawManifest, rawManifestMIMEType)
+		if err != nil {
+			return false, err
+		}
+		remoteDigest, err = list.ChooseInstance(sys)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		remoteDigest, err = manifest.Digest(rawManifest)
+		if err != nil {
+			return false, err
+		}
 	}
 
-	return i.Digest().String() != remoteDigest.String(), nil
+	// Check if we already have that image's manifest in this image.  A
+	// single image can have multiple manifests that describe the same
+	// config blob and layers, so treat any match as a successful match.
+	for _, digest := range append(i.Digests(), i.Digest()) {
+		if digest.Validate() != nil {
+			continue
+		}
+		if digest.String() == remoteDigest.String() {
+			return false, nil
+		}
+	}
+	// No match, but then if we somehow don't have any valid manifests in
+	// this record, consider that a non-match.
+	return len(i.Digests()) == 0 || i.Digest() == "", nil
```

Comment:
What if neither of the two conditions are met.  Wouldn't we return _false_ even if the digests differ?

Reply:
We use the signal that the manifests are different to choose between whether to pull from the reference or not, and in this case I thought we'd want to err on the side of pulling the image.  If you're sure, I can change it.

Reply:
I am slightly more confused.  Let's consider we have a local image `foo` with `digest=1`.

If we compare it to remote image `bar` with `digest=1`, we'll hit line 744 and return false indicating that the digests do not differ.

If we compare the local image to another remote image `hurz` with `digest=2`, we'd hit line 750 and would return false although the digests differ.  But I think we should return true.